src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,437 @@
+            }
+
+            return safeWaitHandles;
+        }
+
+        private static void ReturnSafeWaitHandleArray(SafeWaitHandle[] safeWaitHandles)
+            => t_safeWaitHandlesForRent = safeWaitHandles;
+
+        /// <summary>
+        /// Obtains all of the corresponding safe wait handles and adds a ref to each. Since the <see cref="SafeWaitHandle"/>
+        /// property is publically modifiable, this makes sure that we add and release refs one the same set of safe wait
+        /// handles to keep them alive during a multi-wait operation.
+        /// </summary>
+        private static void ObtainSafeWaitHandles(
+            ReadOnlySpan<WaitHandle> waitHandles,
+            Span<SafeWaitHandle> safeWaitHandles,
+            Span<IntPtr> unsafeWaitHandles)
+        {
+            Debug.Assert(waitHandles != null);
+            Debug.Assert(waitHandles.Length > 0);
+            Debug.Assert(waitHandles.Length <= MaxWaitHandles);
+
+            bool lastSuccess = true;
+            SafeWaitHandle lastSafeWaitHandle = null;
+            try
+            {
+                for (int i = 0; i < waitHandles.Length; ++i)
+                {
+                    WaitHandle waitHandle = waitHandles[i];
+                    if (waitHandle == null)
+                    {
+                        throw new ArgumentNullException("waitHandles[" + i + ']', SR.ArgumentNull_ArrayElement);
+                    }
+
+                    SafeWaitHandle safeWaitHandle = waitHandle._waitHandle;
+                    if (safeWaitHandle == null)
+                    {
+                        // Throw ObjectDisposedException for backward compatibility even though it is not be representative of the issue
+                        throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+                    }
+
+                    lastSafeWaitHandle = safeWaitHandle;
+                    lastSuccess = false;
+                    safeWaitHandle.DangerousAddRef(ref lastSuccess);
+                    safeWaitHandles[i] = safeWaitHandle;
+                    unsafeWaitHandles[i] = safeWaitHandle.DangerousGetHandle();
+                }
+            }
+            catch
+            {
+                for (int i = 0; i < waitHandles.Length; ++i)
+                {
+                    SafeWaitHandle safeWaitHandle = safeWaitHandles[i];
+                    if (safeWaitHandle == null)
+                    {
+                        break;
+                    }
+                    safeWaitHandle.DangerousRelease();
+                    safeWaitHandles[i] = null;
+                    if (safeWaitHandle == lastSafeWaitHandle)
+                    {
+                        lastSafeWaitHandle = null;
+                        lastSuccess = true;
+                    }
+                }
+
+                if (!lastSuccess)
+                {
+                    lastSafeWaitHandle.DangerousRelease();
+                }
+
+                throw;
```

Comment:
Someone had mentioned before that try/catch/throw could cause the stack trace to be lost from the original exception, which is why try/finally was used here instead before. It seems to be working as expected currently on Windows, but not sure if there is an issue with that somewhere. @janvorli do you know?

Reply:
It gets lost if the original `throw` is in the same method. Not sure if it would also happen if some method gets inlined, or if inlining is prevented here.

Reply:
In CoreCLR Windows I'm getting the line number of the original throw even when it's in the same method. In NetFx I'm getting the rethrow line number.